### PR TITLE
fix(scanner): add CA to scanner client library

### PR DESCRIPTION
**Pull request by BradLugo**
## Description

There's currently a bug with `scannerctl` that causes a TLS error when it communicates with `scanner-v4`:
```
{"level":"debug","method":"GetOrCreateImageIndex","component":"scanner/client","image":"quay.io/podman/hello@sha256:1e7354ef111bc9f53fc5fdf4898585d105ce6a7698f7fd03710024dd9775d6f5","rpc":"indexer.GetOrCreateIndexReport","error":"rpc error: code = Unavailable desc = last connection error: connection error: desc = \"transport: authentication handshake failed: x509: “scanner-v4.stackrox” certificate is not trusted\"","duration":1018.714706,"time":"2024-04-09T11:19:05-07:00","message":"retrying gRPC call"}
```

I'm not sure if this is the correct approach (e.g., sensor doesn't have this problem, but I couldn't immediately figure out how sensor configures the CA).

These changes configure the command CA between `scannerctl` and `scanner-v4`.

## Checklist
- ~~[ ] Investigated and inspected CI test results~~
- ~~[ ] Unit test and regression tests added~~
- ~~[ ] Evaluated and added CHANGELOG entry if required~~
- ~~[ ] Determined and documented upgrade steps~~
- ~~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

N/A. `scannerctl` is intended for local development only.

### Here I tell how I validated my change

Before patch:
```
❯ go run ./cmd/scannerctl \
    --certs certs/scannerctl \
    scan --index-only https://quay.io/podman/hello:latest
2024/04/09 11:19:04 auth unspecified: using anonymous auth (use ROX_SCANNERCTL_BASIC_AUTH to set auth)
2024/04/09 11:19:05 image digest: sha256:1e7354ef111bc9f53fc5fdf4898585d105ce6a7698f7fd03710024dd9775d6f5
{"level":"debug","method":"GetOrCreateImageIndex","component":"scanner/client","image":"quay.io/podman/hello@sha256:1e7354ef111bc9f53fc5fdf4898585d105ce6a7698f7fd03710024dd9775d6f5","rpc":"indexer.GetOrCreateIndexReport","error":"rpc error: code = Unavailable desc = last connection error: connection error: desc = \"transport: authentication handshake failed: x509: “scanner-v4.stackrox” certificate is not trusted\"","duration":1018.714706,"time":"2024-04-09T11:19:05-07:00","message":"retrying gRPC call"}
```

After patch:
```
❯ go run ./cmd/scannerctl \
    --certs certs/scannerctl \
    scan --index-only https://quay.io/podman/hello:latest
2024/04/09 11:19:40 auth unspecified: using anonymous auth (use ROX_SCANNERCTL_BASIC_AUTH to set auth)
2024/04/09 11:19:41 image digest: sha256:1e7354ef111bc9f53fc5fdf4898585d105ce6a7698f7fd03710024dd9775d6f5
{
  "hash_id": "/v4/containerimage/sha256:1e7354ef111bc9f53fc5fdf4898585d105ce6a7698f7fd03710024dd9775d6f5",
  "state": "IndexFinished",
  "success": true,
  "contents": {}
}
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by dcaravel — COMMENTED** [review]
Wondering how this impacts when `central` talks to `scanner-v4`.  It appears `central` via the `clientconn` package will automatically [load the CA certs](https://github.com/stackrox/stackrox/blob/be7d4c7c35a1d6be6e75ca23a04ed3fea0215287/pkg/clientconn/client.go#L138) from `/run/secrets/stackrox.io/certs/` when creating the [GRPCConnection](https://github.com/stackrox/stackrox/blob/be7d4c7c35a1d6be6e75ca23a04ed3fea0215287/pkg/clientconn/client.go#L330).
`scannerctl` works fine against `scanner-v4` running in infra by first copying the contents of the `central-tls` secret to `certs/scannerctl` (and setting a port forward):

```sh
files=(ca-key.pem ca.pem cert.pem key.pem) # not all these are needed
for file in "${files[@]}"; do
    k get secrets central-tls -o json | jq -r --arg file "$file" '.data.[$file]' | base64 -d > $file
done
```

```
$ go run ./cmd/scannerctl/ scan --index-only --certs certs/scannerctl https://registry.hub.docker.com/library/hello-world:latest
2024/04/22 16:25:40 auth unspecified: using anonymous auth (use ROX_SCANNERCTL_BASIC_AUTH to set auth)
2024/04/22 16:25:40 image digest: sha256:e2fc4e5012d16e7fe466f5291c476431beaa1f9b90a5c2125b493ed28e2aba57
{
  "hash_id": "/v4/containerimage/sha256:e2fc4e5012d16e7fe466f5291c476431beaa1f9b90a5c2125b493ed28e2aba57",
  "state": "IndexFinished",
  "success": true,
  "contents": {}
}
```
On the surface it looks like this shouldn't impact comms from `central` (IMHO should test this), however I'm curious why/how this makes things work locally for you. Do you by chance have certs sitting in `/run/secrets/stackrox.io/certs/`? Wondering if those are copied to `certs/scannerctl` if the same outcome is achieved?

**Comment by BradLugo** [comment]
> I'm curious why/how this makes things work locally for you. Do you by chance have certs sitting in /run/secrets/stackrox.io/certs/? Wondering if those are copied to certs/scannerctl if the same outcome is achieved?

Yeah, sorry. Now that I've relearned how all this works while answering your question, I realized I should've included a better explanation in the PR description.

Here's my current understanding of the situation:

## Setup & context

`make certs` creates certs for `scanner-v4`, `scanner-v4-db` and `scannerctl` all signed with a common CA. Ultimately, we need to configure this CA _somehow_ so that the client, `scannerctl`, and the server, `scanner-v4` can trust each other. 

Both `scanner-v4` and `scannerctl` have mechanisms to configure their certs + CA:
`scanner-v4`:
https://github.com/stackrox/stackrox/blob/adc5cb1790eb807c41263d438fc8f6c84c7b10d2/scanner/cmd/scanner/main.go#L80-L86
`scannerctl`:
https://github.com/stackrox/stackrox/blob/adc5cb1790eb807c41263d438fc8f6c84c7b10d2/scanner/cmd/scannerctl/main.go#L77-L91

Now that I've typed the above out, I realize I should've included the following info in the description:
I first ran `make certs` to generate the certs. Then, ran `scanner-v4` with the following config:
```
indexer:
  database:
    conn_string: <redacted>
matcher:
  enable: false
mtls:
  certs_dir: certs/scanner-v4
log_level: debug
```
And then, as noted in the description already, ran `scannerctl` with the `--certs` flag.

## scanner-v4 TLS

IIUC, we look at `/run/secrets/stackrox.io/certs` by default but then override it with the yaml config, if it's specified. In my case, it is, so we need `scannerctl` to also configure its certs with the common CA correctly.

Here's the evidence I found:

### TL;DR
Here's where we first start configuring TLS: https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/pkg/grpc/endpoints.go#L151-L157
Which eventually calls this: https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/pkg/env/setting.go#L87-L101

### Long version
<details>

<summary>Call stack in reverse order (call queue?)</summary>

https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/scanner/cmd/scanner/main.go#L118
https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/scanner/cmd/scanner/main.go#L118
https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/pkg/grpc/endpoints.go#L153
https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/pkg/mtls/verifier/verify.go#L58
https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/pkg/mtls/crypto.go#L141
https://github.com/stackrox/stackrox/blob/bd072a68db3b0a7cf76de8eeeed747fe24e70d6c/pkg/env/setting.go#L87-L101

</details>

## The problem - scannerctl TLS

AFAICT, we're not configuring the CA properly in `scannerctl`. I tried debugging this again while writing this response, and I'm still stumped as to why it doesn't work. I would expect this to take care of the CA: https://github.com/stackrox/stackrox/blob/1a788a35c2c4e2802327e60eaaa235e80536c43b/pkg/clientconn/client.go#L133-L151
But for whatever reason, it doesn't, and I found `AddRootCAs` to set the RootCA field and moves the needle: https://github.com/stackrox/stackrox/blob/53424d947392007e96c2403b8786564affaf7da9/pkg/mtls/crypto.go#L170

---

With all that being said, I'm not sure this is the correct way the fix the problem, especially since I don't fully understand it. There are a few too many levels of abstraction for me to fully wrap my head around it, and I suspect this is already solved in other places (e.g., maybe sensor's connection to scanner-v4? - not sure). If anyone has any pointers on this, please let me know

File: pkg/scannerv4/client/client.go [hunk 1]
```diff
@@ -138,6 +138,7 @@ func createGRPCConn(ctx context.Context, o connOptions) (*grpc.ClientConn, error
 		clientconn.ServerName(o.serverName),
 		clientconn.MaxMsgReceiveSize(maxRespMsgSize),
 		clientconn.WithDialOptions(dialOpts...),
+		clientconn.AddRootCAs(),
```
Comment (RTann) [inline pkg/scannerv4/client/client.go:141]: It doesn't seem like this should be needed. How are you running this? Completely locally or deployed via Helm?

I don't understand how this fixes the problem (I believe that it does for you, but I'm confused as to why).

If we do end up adding this, we should do it conditionally, so it doesn't affect prod deployments. From what I can tell, they've been working as-is

Comment (BradLugo) [reply pkg/scannerv4/client/client.go:141]: > How are you running this? Completely locally or deployed via Helm?
Completely locally

> I don't understand how this fixes the problem

I explained here: https://github.com/stackrox/stackrox/pull/10687#issuecomment-2075713272
The TL;DR is `AddRootCAs` will add the CA we configure in the environment variables. I can find which paritcular ones but it's all or some subset of these:
https://github.com/stackrox/stackrox/blob/44e3d1daf95517e70468af2de6bd9d62d937e393/scanner/cmd/scannerctl/main.go#L78-L90

> If we do end up adding this, we should do it conditionally, so it doesn't affect prod deployments. From what I can tell, they've been working as-is
Yeah, I definitely not sure if this is the correct way to fix the problem. In fact, after speaking with Dave, I'm wondering why some of our TLS verification/management works the way it does in general

Comment (RTann) [reply pkg/scannerv4/client/client.go:141]: I finally looked into this again, and I pretty much came to this same conclusion as you. Have you also tried running this in k8s or openshift to see if it still works in a prod environment?

Comment (RTann) [reply pkg/scannerv4/client/client.go:141]: I'm actually still confused, though, as, without this, I believe we use the system's CA certs. I believe Central is typically the CA, and I don't see where we add its cert into our system's CA cert pool. Do you know where/if we do that?
